hw/xbox/nv2a/nv2a_pgraph.c

```diff
@@ -3083,6 +3135,28 @@ static void pgraph_bind_shaders(PGRAPHState *pg)
 
     glUseProgram(pg->shader_binding->gl_program);
 
+    /* Clipping regions */
+    for (i = 0; i < state.psh.window_clip_count; i++) {
+        char name[32];
+
+        uint32_t x = pg->regs[NV_PGRAPH_WINDOWCLIPX0 + i * 4];
+        GLuint xmin = GET_MASK(x, NV_PGRAPH_WINDOWCLIPX0_XMIN);
+        GLuint xmax = GET_MASK(x, NV_PGRAPH_WINDOWCLIPX0_XMAX);
+        uint32_t y = pg->regs[NV_PGRAPH_WINDOWCLIPY0 + i * 4];
+        GLuint ymin = pg->surface_shape.clip_height - 1 -
+                      GET_MASK(y, NV_PGRAPH_WINDOWCLIPX0_XMIN);
+        GLuint ymax = pg->surface_shape.clip_height - 1 -
+                      GET_MASK(y, NV_PGRAPH_WINDOWCLIPX0_XMAX);
```

Comment:
nit: These inversions are necessary for GL / D3D differences. I think it would be better to do these in 2 steps:

```c
GLuint ymin = GET_MASK(y, NV_PGRAPH_WINDOWCLIPX0_XMIN);
GLuint ymax = GET_MASK(y, NV_PGRAPH_WINDOWCLIPX0_XMAX);

// <comment about flipping coordinate system upside down as we'll use GL coordinates in shader, which work different from D3D>
ymin = pg->surface_shape.clip_height - 1 - ymin;
ymax = pg->surface_shape.clip_height - 1 - ymax;
```

Reply:
What's strange about this remark, it's that in most cases NV2A hardware is closely mapped to OpenGL. So why are the clipping coordinates following the Direct3D coordinate system all of a sudden?